Implemented Player Manager

Platnumrcn14 opened the pull request:
-Manages AI players and USER differentiation in PlayerType -Manages if the player is IN (waiting to bet), FOLDED, CALLED(has matched high bet and ready for next round) or ALLIN(out of fund and unable to bet further) in PlayerStatus -Manages if the player is DEALER, SMALLBLIND,  BIGBLIND, or NONE in PlayerPosition -Manages player funds in Texas Holdem in PlayerStack -Manages the bet the player has placed in the current round in PlayerBet
![2024-11-19 (9)](https://github.com/user-attachments/assets/dd76b8c7-5db6-4b38-b900-7e95cd816f70)
![2024-11-19 (10)](https://github.com/user-attachments/assets/26e8e7ad-27bf-4711-a4cf-b93dc4254bd1)

-CreateUser() creates the user and initiates the amount of funds (Set to 200 as a default can be modified if needed)
-CreateAIPlayers(int numPlayers) adds AI players to the Players class and initiates funds (Set to 200 as a default) 
-int GetPreflopIndex() returns the position of the player who will have the first turn to bet at the beginning of the round (always player to the left of the big blind) 
![2024-11-19 (11)](https://github.com/user-attachments/assets/59da4bdd-76d3-4df1-95bf-7d54ef999793)

-int GetStartingBettorIndex() will start with the index of the small blind and check each player's status until the first player is found that is not FOLDED or ALLIN, this player will be the first to bet. 
-bool IsActivePlayer(int playerIndex) will check if the player at the selected index is eligible to bet in the round 
-Call(int amount, int playerIndex) will excecute a call bet to match the current bet placed by another player, subtract the amount from the player's stack and set the player as ready to advance to the next round 
![2024-11-19 (12)](https://github.com/user-attachments/assets/25d529ec-e6b6-40ca-a61f-b64c9f0d16af)

-Raise(int amount, int playerIndex) will excecute the same logic as call, however since the bet is higher it will reset the status of all players with the status CALLED to IN to indicate they must match the new current bet before continuing 
-int AllInBet(int playerIndex) will shift all of the players funds as their bet and set their status to ALLIN, and return the amount the player has bet as the new current bet 
-Fold(int playerIndex) sets the players status to FOLDED, they will no longer be participating in the round 
-Check(int playerIndex) will set the player's status as CALLED (ready for the next round) 
![2024-11-19 (13)](https://github.com/user-attachments/assets/2ed00932-24c4-4c44-9d2a-20a471bdf820)

-bool AdvanceRound() will still need to be implemented, it will contain the conditions needed to advance to the next round. 
-ResetBets() will be called once all player's bets have been added to the pot, it will reset all player bets to 0 for the next round of betting 
-List<int> PackageFoldedBets() will package the bets of folded players onto a list to be added to the pot. This is an important step in case player(s) decide to raise, then are raised back, then decide to fold. This way the bets are not mistaken for all in bets by the pot manager 
![2024-11-19 (14)](https://github.com/user-attachments/assets/ede5bc71-35d5-4908-8b14-c2efd7e29f75)

-List<int> PackageBets() at the end of a round will package the bets of players that have called or have just gone all-in to be added to the pot and create side pots as needed 
-List<int> GetActivePlayersIndexes() will return the list of players that are eligible for the current pot. 
-EliminatePlayers() will check if a player's stack is 0, if it is they are removed from the table 
![2024-11-19 (15)](https://github.com/user-attachments/assets/9efdb886-bff5-44ba-a09d-4d29c6403c2e)

-InitiateBlinds() will randomly select which player is the DEALER at the beginning of the game and position the SMALLBLIND and BIGBLIND to the dealer's left 
-SetNextRoundBlinds() will shift the blinds and the dealer one position to the left each, if only two players remain there is no longer a dealer only a small and big blind 
![2024-11-19 (16)](https://github.com/user-attachments/assets/0779dc5b-3a05-45f6-a54f-0b9f3e2139a5)

-GetBlinds(int _smallBlind, int _bigBlind) will place the blind bets for the appropriate players , if the blinds are larger than that player's stack, that player will automatically be all-in
![2024-11-19 (17)](https://github.com/user-attachments/assets/10314adb-e597-4f70-adbd-06bfbaecd2dc)

Platnumrcn14 committed:
Implemented Player Manager
-Manages AI players and USER differentiation in PlayerType
-Manages if the player is IN (waiting to bet), FOLDED, CALLED(has matched high bet and ready for next round) or ALLIN(out of fund and unable to bet further) in PlayerStatus
-Manages if the player is DEALER, SMALLBLIND,  BIGBLIND, or NONE in PlayerPosition
-Manages player funds in Texas Holdem in PlayerStack
-Manages the bet the player has placed in the current round in PlayerBet

-CreateUser() creates the user and initiates the amount of funds (Set to 200 as a default can be modified if needed)
-CreateAIPlayers(int numPlayers) adds AI players to the Players class and initiates funds (Set to 200 as a default)
-int GetPreflopIndex() returns the position of the player who will have the first turn to bet at the beginning of the round (always player to the left of the big blind)
-int GetStartingBettorIndex() will start with the index of the small blind and check each player's status until the first player is found that is not FOLDED or ALLIN, this player will be the first to bet.
-bool IsActivePlayer(int playerIndex) will check if the player at the selected index is eligible to bet in the round
-Call(int amount, int playerIndex) will excecute a call bet to match the current bet placed by another player, subtract the amount from the player's stack and set the player as ready to advance to the next round
-Raise(int amount, int playerIndex) will excecute the same logic as call, however since the bet is higher it will reset the status of all players with the status CALLED to IN to indicate they must match the new current bet before continuing
-int AllInBet(int playerIndex) will shift all of the players funds as their bet and set their status to ALLIN, and return the amount the player has bet as the new current bet
-Fold(int playerIndex) sets the players status to FOLDED, they will no longer be participating in the round
-Check(int playerIndex) will set the player's status as CALLED (ready for the next round)
-bool AdvanceRound() will still need to be implemented, it will contain the conditions needed to advance to the next round.
-ResetBets() will be called once all player's bets have been added to the pot, it will reset all player bets to 0 for the next round of betting
-List<int> PackageFoldedBets() will package the bets of folded players onto a list to be added to the pot. This is an important step in case player(s) decide to raise, then are raised back, then decide to fold. This way the bets are not mistaken for all in bets by the pot manager
-List<int> PackageBets() at the end of a round will package the bets of players that have called or have just gone all-in to be added to the pot and create side pots as needed
-List<int> GetActivePlayersIndexes() will return the list of players that are eligible for the current pot.
-EliminatePlayers() will check if a player's stack is 0, if it is they are removed from the table
-InitiateBlinds() will randomly select which player is the DEALER at the beginning of the game and position the SMALLBLIND and BIGBLIND to the dealer's left
-SetNextRoundBlinds() will shift the blinds and the dealer one position to the left each, if only two players remain there is no longer a dealer only a small and big blind
-GetBlinds(int _smallBlind, int _bigBlind) will place the blind bets for the appropriate players , if the blinds are larger than that player's stack, that player will automatically be all-in

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +825,329 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
             timer.Stop();
             timer.Dispose();
         }
+        public class PlayerManager
```
Rename to HoldEmPlayerManager or PokerPlayerManager.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +825,329 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
         }
+        public class PlayerManager
+        {
+            // List to hold the pots for the game
```
These need to be summary comments, not regular comments.

Use the "/// <summary>" format.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +825,329 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+
+            public PlayerManager()
+            {
+                Players = new List<Player>(); // Initialize the Players list
```
This can be initialized where Players is defined, like "{get; set;} = new();". Avoids the need for a constructor.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +825,329 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            /// <summary>
+            /// Manages the pot in a Texas Hold'em game, including adding and distributing chips.
+            /// </summary>
+            public class Player
```
Classes should only be double nested like this when they are only used in the above class. This should not be the case.

Platnumrcn14 committed:
-Refactored and added summary statements
-Added EligiblePlayers list return function to return a list of eligible players for pot distribution purposes in round conclusion
-Added logic for funds distribution at the end of a round
-Added boolean methods in Players class to determine if the game should advance to the end of the round

TO-DOs
-UnitTest cases for logic
-Peer reviewed documentation edits & any logic modifications

modmo22 reviewed (approved):
The only real feedback I have of this at this point is to maybe consider putting the player manager in its own module. I'd like some input from the rest of the team about it, though. Jacob has the better head for modular organization compared to me. I don't see anything alarming. Once some unit tests are done, this should be pretty solid

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -25,6 +25,7 @@
 using Microsoft.Xna.Framework.Content;
 using Microsoft.Xna.Framework.Graphics;
 using Microsoft.Xna.Framework.Input;
+using static CardsCashCasino.Manager.PlayerManager;
```
manager classes should never be static. The C# convention says: managers are classes, utils are static objects. Just make an instance of PlayerManager in the class.

Conversely, if you want to statically reference something NEVER do this. Always call PlayerManager.XXXX.

If this can be static and does not need to be a class, then make it a Util class.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -48,6 +52,36 @@ public enum BettingPhase
+        SIDE
+    }
+
+    public enum PlayerType
```
we need some comments on these enums.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -448,54 +511,79 @@ private Point GetNewCursorPos()
-            // _blindIncreaseCountdown = 5;
+
+            //set initial blinds and ante, set countdown to increase blinds after 4 rounds of play
+            _ante = 2;
```
Could you add a comment being a little more descriptive as to why these values are being set this way? I assume its a texas hold em specific thing but it will make any bug fixing later on down the road WAAAY easier.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -583,80 +666,104 @@ public void DealRiver()
-                Tuple<List<Card>, PokerUtil.Ranking> pair = PokerUtil.GetScore(_communityCards, _playerHands[i].Cards.ToList());
-                // If this ties the best ranking
-                if (pair.Item2 == bestRanking)
+                eligiblePlayers = _potManager.PlayersEligible(potNumber);
```
confused by this line. we define the list of eligible players before the for loop, then immediately reassign it here. is there a reason we need to ever initialize it before the for loop?

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -665,74 +772,249 @@ private void HandleBettingPhase()
+            {
+                return;
+            }
+            int playerIndex;
```
delete this. you are just defining it below. do it there only.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -665,74 +772,249 @@ private void HandleBettingPhase()
+        /// <summary>
+        /// handles the initial stage of betting before any community cards appear
+        /// </summary>
         private void HandlePreflop()
```
I am very confused what is happening here and I think it will not be able to be integrated into the update loop.

Is the intention that all of this code happens at once in a single game frame? if so, it is fine BUT if it is intended that this is a broad method to handle multiple game frames or parts of the game, i.e., each player doing an action, this will NOT work because you are nesting a while loop within the main Update loop. Instead of this, you should organize this so that the ONLY loop is the Update loop.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -665,74 +772,249 @@ private void HandleBettingPhase()
+            playerIndex = _players.GetStartingBettorIndex();
+
+            // Set the player index to the small blind or the player closest to the small blind
+            while (!_players.AdvanceRound())
```
The same question as above comes here. We are nesting a while loop within the main Update loop, which we cannot do. That will just stall the game. Refactor this and the above to be within the Update loop.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -665,74 +772,249 @@ private void HandleBettingPhase()
+            {
+                return;
+            }
+            int playerIndex;
```
remove this, define it below like you already do.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -665,74 +772,249 @@ private void HandleBettingPhase()
+                return;
+            }
+            //initiate player index and reset the current round bet to 0
+            int playerIndex;
```
again, get rid of this.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -665,74 +772,249 @@ private void HandleBettingPhase()
+            playerIndex = _players.GetStartingBettorIndex();
+
+            // iterate through the players starting with the first player in and handle their actions.
+            while (!_players.AdvanceRound())
```
this will also need to be refactored.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -665,74 +772,249 @@ private void HandleBettingPhase()
+            playerIndex = _players.GetStartingBettorIndex();
+
+            // Set the player index to the small blind or the player closest to the small blind
+            while (!_players.AdvanceRound() && !_players.OnePlayerLeft())
```
this will also need to be refactored.

jacobwilki01 reviewed (dismissed):
I am concerned this code will not be able to be integrated into the Update loop. Please see the comments.

Platnumrcn14 committed:
Completes tickets 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.2.6, and 2.4.2
-Implemented GenerateAntes(int _ante) function adding the antes to the pot from every player
-Removed InititeBlinds() function and instead nested the logic into InitiatePlayers(int numAIs) function
-Corrrected logic flaws in Call(), ResetBets(), PackageFoldedBets(), PackageBets(), SetNextRoundBlinds(), CollectBlinds(), and CreateSidePots() functions
-Temporarially set PlayerStack default value to 500, will be tied to player's overall funds

modmo22 committed:
Added missing XML comments.

Platnumrcn14 committed:
-Refactored 'while' statements into 'if' statements
-corrected other notes

Platnumrcn14 committed:
-refactored 'while' statements into 'if' statement format
-corrected other errors present

eBerkley committed:
Added code to determine user's action

eBerkley committed:
Forgot to reset _userRaising

eBerkley committed:
Added some `RoundLogic()` code

eBerkley committed:
Forgot to add code that calls RoundLogic()

eBerkley committed:
Added a NextPhase() call

Platnumrcn14 committed:
-finished review comments

modmo22 committed:
Eliminated unneeded code

modmo22 committed:
Merge branch 'develop' into feature-players-manager

modmo22 committed:
Merge remote-tracking branch 'origin/feature-players-manager' into feature-players-manager

eBerkley committed:
Started to get hold em to render, has a bug related to deck size atm

eBerkley committed:
Added placeholders

eBerkley committed:
Fixed a issue

modmo22 committed:
QOL tweaks to HoldEmTests.cs

eBerkley reviewed (approved):
Good enough for now
